File: src/engine/sprite/sprite_serialize.cc
// Copyright 2025 m4jr0. All Rights Reserved.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

// Precompiled. ////////////////////////////////////////////////////////////////
#include "precompiled.h"
////////////////////////////////////////////////////////////////////////////////

// Header. /////////////////////////////////////////////////////////////////////
#include "sprite_serialize.h"
////////////////////////////////////////////////////////////////////////////////

namespace rl {
void writeSprite(std::ostream& os, const Sprite& s) {
  writePod(os, s.rect);
  writePod(os, s.offset);
  writePod(os, s.offsetFlipped);
  writePod(os, s.sourceSize);
}

void readSprite(std::istream& is, Sprite& s) {
  readPod(is, s.rect);
  readPod(is, s.offset);
  readPod(is, s.offsetFlipped);
  readPod(is, s.sourceSize);
}

void writeAtlasResource(std::ostream& os, const AtlasResource& a) {
  writePod(os, a.id);
  writePod(os, a.tex);

  auto count = a.sprites.size();
  writePod(os, count);

  for (const auto& [name, idx] : a.nameToIdx) {
    const auto& s = a.sprites[idx];

    auto nameLen = static_cast<u32>(name.size());
    writePod(os, nameLen);
    os.write(name.data(), static_cast<std::streamsize>(name.size()));

    writeSprite(os, s);
  }
}

void readAtlasResource(std::istream& is, AtlasResource& a) {
  readPod(is, a.id);
  readPod(is, a.tex);

  usize count;
  readPod(is, count);

  a.sprites.resize(count);
  a.nameToIdx.clear();
  a.nameToIdx.reserve(count);

  for (usize i = 0; i < count; ++i) {
    u32 nameLen;
    readPod(is, nameLen);

    std::string name(nameLen, '\0');
    is.read(name.data(), static_cast<std::streamsize>(nameLen));

    auto idx = static_cast<SpriteIndex>(i);
    a.nameToIdx.emplace(std::move(name), idx);

    readSprite(is, a.sprites[i]);
  }
}
}  // namespace rl

File: tools/rlres/exporter/engine/sprite/sprite_serialize.py
# Copyright 2025 m4jr0. All Rights Reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

from __future__ import annotations

import struct

from typing import BinaryIO

from rlres.data.engine.sprite.sprite_type import Sprite


_RECT_PACK = struct.Struct("<iiii")
_OFFSET_PACK = struct.Struct("<ff")
_SOURCE_SIZE_PACK = struct.Struct("<ii")


def write_sprite(f: BinaryIO, s: Sprite) -> None:
    f.write(_RECT_PACK.pack(s.rect.x, s.rect.y, s.rect.w, s.rect.h))
    f.write(_OFFSET_PACK.pack(*s.offset))
    f.write(_OFFSET_PACK.pack(*s.offset_flipped))
    f.write(_SOURCE_SIZE_PACK.pack(*s.source_size))
